x-pack/plugins/ml/public/application/model_management/get_model_state.tsx

```diff
@@ -5,13 +5,28 @@
  * 2.0.
  */
 
-import type { ModelState } from '@kbn/ml-trained-models-utils';
-import { MODEL_STATE } from '@kbn/ml-trained-models-utils';
+import { DEPLOYMENT_STATE, MODEL_STATE, type ModelState } from '@kbn/ml-trained-models-utils';
 import type { EuiHealthProps } from '@elastic/eui';
 import { i18n } from '@kbn/i18n';
+import type { ModelItem } from './models_list';
+
+/**
+ * Resolves result model state based on the download status and deployments stats.
+ */
+export const getModelDeploymentState = (model: ModelItem): ModelState | undefined => {
```

Comment:
It would be nicer if this could use a single loop rather than three. But I appreciate that the code would be more complicated where as this is simple to read.

Reply:
Single loop solutions would look something like this:
```typescript
export const getModelDeploymentState = (model: ModelItem): ModelState | undefined => {
  let modelState: ModelState | undefined;
  let isStarting = false;
  let isStopping = false;
  for (const deployment of model.stats?.deployment_stats ?? []) {
    if (deployment.state === DEPLOYMENT_STATE.STARTED) {
      return MODEL_STATE.STARTED;
    }
    if (deployment.state === DEPLOYMENT_STATE.STARTING) {
      isStarting = true;
    }
    if (!isStarting && deployment.state === DEPLOYMENT_STATE.STOPPING) {
      isStopping = true;
    }
    modelState = deployment.state;
  }
  return isStarting ? MODEL_STATE.STARTING : isStopping ? MODEL_STATE.STOPPING : modelState;
};
```
I reckon array methods are easier to follow.

Reply:
Agreed